### PR TITLE
GridTools::MarchingCubeAlgorithm: merge duplicate vertices in process()

### PR DESCRIPTION
**Pull request by mschreter**
With this PR the determined vertices by means of the MCA are written as unique values. 

Follow-up to and ~~depends on #14305~~.

@peterrum: So far I only use it inside the `process()`-function that returns vertices.

### CODE REVIEW

File: source/grid/grid_tools.cc [hunk 1]
```diff
@@ -6647,6 +6648,136 @@ namespace GridTools
...
+                  n_vertices_old;
+            }
+        }
+    }
+
+
+
+    template <int dim>
+    void
+    merge_duplicate_points(std::vector<Point<dim>> &vertices)
```
Comment (mschreter) [inline source/grid/grid_tools.cc:6752]: I had to define this Function prior to the `process` function so I also moved the `internal::process_sub_cell()` function.

**Review by peterrum — APPROVED** [review]

**Comment by peterrum** [comment]
> If we really want to use it also in case we request CellData<dim>, we need to change this part.

Exactly. Renumber the vertices of the cells. You already have this info, don't you?

File: source/grid/grid_tools.cc (hunk 1 above)
Comment (mschreter) [reply source/grid/grid_tools.cc]: > Exactly. Renumber the vertices of the cells. You already have this info, don't you? You could use the same function; if it gets an empty vector of cells, it does nothing.

I tried to include this, see https://github.com/dealii/dealii/pull/14323/commits/587ccb50d3553c9a4ea4f3d724b89c500a146b11. I am not sure about the performance of accessing the `Point<dim>` key of the map for every cell-vertex-point. If I remap the cell vertices, as soon as `create_triangulation` is called, an assert is thrown

```cpp
--------------------------------------------------------
An error occurred in line <11715> of file </home/magdalena/TUM/dealii_work/source/grid/tria.cc> in function
    void dealii::Triangulation<<anonymous>, <anonymous> >::create_triangulation(const std::vector<dealii::Point<spacedim> >&, const std::vector<dealii::CellData<dim> >&, const dealii::SubCellData&) [with int dim = 1; int spacedim = 2]
The violated condition was: 
    !(correct(i, j) ^ (neighbor->direction_flag() == (*cell)->direction_flag()))
Additional information: 
    (none)
--------------------------------------------------------
```
Do you know by chance what we need to consider here regarding the vertex numbering?

Comment (peterrum) [reply source/grid/grid_tools.cc]: > [with int dim = 1; int spacedim = 2]

It is a 1D mesh in 2D space, i.e., we have simple lines here and the resulting mesh is not orientatble. This is odd. I don't know when this is the case. Does anyone know and/or have a solution?

For the time being, let's not merge the vertices in the `CellData` case.

Comment (mschreter) [reply source/grid/grid_tools.cc]: done!

**Review by peterrum — APPROVED** [review]

File: source/grid/grid_tools.cc [hunk 2]
```diff
@@ -867,6 +868,37 @@ namespace GridTools
...
+  void
+  delete_duplicated_vertices(std::vector<Point<dim>> &vertices,
+                             const double             tol)
+  {
+    if (vertices.size() == 0)
+      return;
+
+    // 1) map point to local vertex index
+    std::map<Point<dim>, unsigned int, FloatingPointComparator<double>>
+      map_point_to_local_vertex_index{FloatingPointComparator<double>(tol)};
```
Comment (mschreter) [inline source/grid/grid_tools.cc:881]: @peterrum: I needed the curly braces to get it compiled!

**Comment by mschreter** [comment]
@peterrum how shall we proceed here?

**Review by kronbichler — APPROVED** [review]

**Pull request merged**